Examples/Io/Root/src/RootDigitizationWriter.cpp

```diff
@@ -0,0 +1,169 @@
+      std::move(dTrees));
+}
+
+ActsExamples::RootDigitizationWriter::~RootDigitizationWriter() {
+  /// Close the file if it's yours
+  if (m_cfg.rootFile == nullptr) {
+    m_outputFile->cd();
+    for (auto dTree = m_outputTrees.begin(); dTree != m_outputTrees.end();
+         ++dTree) {
+      (*dTree)->tree->Write();
+    }
+    m_outputFile->Close();
+  }
+}
+
+ActsExamples::ProcessCode ActsExamples::RootDigitizationWriter::endRun() {
+  // Write the tree
+  return ProcessCode::SUCCESS;
+}
+
+ActsExamples::ProcessCode ActsExamples::RootDigitizationWriter::writeT(
+    const AlgorithmContext& ctx,
+    const ActsExamples::GeometryIdMultimap<
+        Acts::FittableMeasurement<ActsExamples::DigitizedHit>>& measurements) {
+  // composition
+
+  const auto& simHitContainer =
+      ctx.eventStore.get<SimHitContainer>(m_cfg.inputSimulatedHits);
+
+  for (auto& [key, value] : measurements) {
+    std::visit(
+        [&](auto&& m) {
+          const auto& surface = m.referenceObject();
+          Acts::GeometryIdentifier sIdentifier = surface.geometryId();
+          auto dTreeItr = m_outputTrees.find(sIdentifier);
+          if (dTreeItr != m_outputTrees.end()) {
+            auto& dTree = *dTreeItr;
+            // Fill the identification
+            dTree->fillIdentification(ctx.eventNumber, sIdentifier);
+            auto hitIndices = m.sourceLink().hitIndices();
+            std::vector<SimHit> simHits;
+            simHits.reserve(hitIndices.size());
+            for (auto& hi : hitIndices) {
+              auto nthhit = simHitContainer.nth(hi);
+              simHits.push_back(*nthhit);
+            }
+
+            auto tParams = truthParameters(ctx.geoContext, surface, simHits);
+            dTree->fillTruthParameters(std::get<0>(tParams),
+                                       std::get<1>(tParams),
+                                       std::get<2>(tParams));
+            dTree->fillBoundMeasurement(m);
+            dTree->tree->Fill();
+          }
+        },
+        value);
+  }
+
+  // Exclusive access to the tree while writing
+  std::lock_guard<std::mutex> lock(m_writeMutex);
+
+  return ActsExamples::ProcessCode::SUCCESS;
+}
+
+std::tuple<Acts::Vector2D, Acts::Vector4D, Acts::Vector3D>
+ActsExamples::RootDigitizationWriter::truthParameters(
+    const Acts::GeometryContext& gCtx, const Acts::Surface& surface,
+    const std::vector<ActsExamples::SimHit>& simulatedHits) {
+  Acts::Vector3D position(0., 0., 0.);
+  Acts::Vector3D direction(0., 0., 0.);
+  double ctime = 0.;
+
+  if (simulatedHits.size() == 1 and
+      surface.isOnSurface(gCtx, simulatedHits[0].position(),
+                          simulatedHits[0].unitDirection())) {
+    position = simulatedHits[0].position();
+    direction = simulatedHits[0].unitDirection();
+    ctime = simulatedHits[0].time();
+
+  } else if (simulatedHits.size() > 1) {
+    Acts::Vector4D avePos(0., 0., 0., 0.);
+    if (simulatedHits.size() > 1) {
```

Comment:
Why do we need to check the size again here since it’s already done earlier?

Reply:
Artifact, removed.